pkg/virtualfs/directory.go

```diff
@@ -0,0 +1,169 @@
+	readdirError error
+	onReaddir    func()
+}
+
+// Summary returns summary of a directory.
+func (d *Directory) Summary() *fs.DirectorySummary {
+	return nil
+}
+
+// AddFileWithStreamSource adds a mock file with the specified name, permissions and source.
+func (d *Directory) AddFileWithStreamSource(name, sourceEndpoint string, permissions os.FileMode) (file *File, err error) {
+	d, name, err = d.resolveAndAddSubdir(name)
+	if err != nil {
+		return nil, err
+	}
+
+	file = &File{
+		dirEntry: dirEntry{
+			name: name,
+			mode: permissions,
+			// TODO: Add owner information
+		},
+		source: func() (ReaderSeekerCloser, error) {
+			return streamReader(sourceEndpoint)
+		},
+	}
+
+	err = d.addChild(file)
+
+	return file, errors.Wrap(err, "Failed to add file")
+}
+
+// AddDir adds a fake directory with a given name and permissions.
+func (d *Directory) AddDir(name string, permissions os.FileMode) (subdir *Directory, err error) {
+	d, name, err = d.resolveSubdir(name)
+	if err != nil {
+		return nil, err
+	}
+
+	subdir = &Directory{
+		dirEntry: dirEntry{
+			name: name,
+			mode: permissions | os.ModeDir,
+		},
+	}
+
+	err = d.addChild(subdir)
+
+	return subdir, err
+}
+
+// Subdir finds a subdirectory with the given name.
+func (d *Directory) Subdir(name ...string) (*Directory, error) {
+	i := d
+
+	for _, n := range name {
+		i2 := i.children.FindByName(n)
+		if i2 == nil {
+			return nil, errors.New(fmt.Sprintf("'%s' not found in '%s'", n, i.Name()))
+		}
+
+		if !i2.IsDir() {
+			return nil, errors.New(fmt.Sprintf("'%s' is not a directory in '%s'", n, i.Name()))
+		}
+
+		i = i2.(*Directory)
+	}
+
+	return i, nil
+}
+
+// Remove removes directory dirEntry with the given name.
+func (d *Directory) Remove(name string) {
+	newChildren := d.children[:0]
+
+	for _, e := range d.children {
+		if e.Name() != name {
+			newChildren = append(newChildren, e)
+		}
+	}
+
+	d.children = newChildren
+}
+
+// OnReaddir invokes the provided function on read.
+func (d *Directory) OnReaddir(cb func()) {
+	d.onReaddir = cb
+}
+
+// Child gets the named child of a directory.
+func (d *Directory) Child(ctx context.Context, name string) (fs.Entry, error) {
+	return fs.ReadDirAndFindChild(ctx, d, name)
+}
+
+// Readdir gets the contents of a directory.
+func (d *Directory) Readdir(ctx context.Context) (fs.Entries, error) {
+	if d.readdirError != nil {
+		return nil, d.readdirError
+	}
+
+	if d.onReaddir != nil {
+		d.onReaddir()
+	}
+
+	return append(fs.Entries(nil), d.children...), nil
+}
+
+func (d *Directory) addChild(e fs.Entry) error {
+	if strings.Contains(e.Name(), "/") {
+		return errors.New("Failed to add child entry: name cannot contain '/'")
+	}
+
+	d.children = append(d.children, e)
+	d.children.Sort()
+	return nil
+}
+
+func (d *Directory) resolveSubdir(name string) (parent *Directory, leaf string, err error) {
+	parts := strings.Split(name, "/")
+	for _, n := range parts[0 : len(parts)-1] {
+		if d, err = d.Subdir(n); err != nil {
```

Comment:
Do not override the receiver variable (d)! May work but its not a normal thing to do and any linter would complain!